### PR TITLE
boot: helpers for manipulating current and good recovery systems list

### PR DESCRIPTION
**Pull request by bboozzoo**
This should be the last batch of helpers for manipulating the lists of current and good (i.e. tested) recovery systems.

Based on top of #10252, the relevant commit is https://github.com/snapcore/snapd/commit/0d660a129a522027d7c314befd922e66f9a661f4

### CODE REVIEW

**Comment by bboozzoo** [comment]
Hm looks like I missed some bits when splitting up patches. Marking as blocked for now.

File: boot/systems.go [hunk 1]
```diff
@@ -342,3 +343,93 @@ func InspectTryRecoverySystemOutcome(dev Device) (outcome TryRecoverySystemOutco
...
+func MarkRecoverySystemGood(dev Device, systemLabel string) (err error) {
+	if !dev.HasModeenv() {
+		return fmt.Errorf("internal error: recovery systems can only be used on UC20")
+	}
+
+	m, err := loadModeenv()
+	if err != nil {
+		return err
+	}
+	if !strutil.ListContains(m.CurrentRecoverySystems, systemLabel) {
```
Comment (pedronis) [inline boot/systems.go:362]: does this check make sense,  afair the code we added in devicemgr.go wil clear the tried label out of this list?  so I'm entirely sure how this is expected to be used, will something add it back first?

**Review by pedronis — COMMENTED** [review]
a question

File: boot/systems.go (hunk 1 above)
Comment (bboozzoo) [reply boot/systems.go:362]: No, you're right. Also we rewrite the modeenv using AtomicFile so in theory we should have both entries or none. This is how it's used, at least in the non-remodel scenario.

File: boot/systems.go [hunk 2]
```diff
@@ -342,3 +343,99 @@ func InspectTryRecoverySystemOutcome(dev Device) (outcome TryRecoverySystemOutco
 
 	return outcome, trySystem, nil
 }
+
+// MarkRecoverySystemGood marks the provided recovery system as a good one, and
+// esures that the system is present in the list of good recovery systems and
+// current recovery systems in modeenv. If the system uses encryption, the keys
+// will updated state. If resealing fails, an attempt to restore the previous
+// state is made
+func MarkRecoverySystemGood(dev Device, systemLabel string) (err error) {
```
Comment (pedronis) [inline boot/systems.go:352]: we discussed a different name and signature for this in the context of the next PR

Comment (bboozzoo) [reply boot/systems.go]: The API has been tweaked as discussed.

**Review by pedronis — APPROVED** [review]
thank you

File: boot/systems.go (hunk 1 above)
Comment (mvo5) [reply boot/systems.go:368]: Hm, was this addressed and the conversation is stale?

File: boot/systems.go [hunk 3]
```diff
@@ -342,3 +343,84 @@ func InspectTryRecoverySystemOutcome(dev Device) (outcome TryRecoverySystemOutco
...
+		m.GoodRecoverySystems = append(m.GoodRecoverySystems, systemLabel)
+		rewriteModeenv = true
+	}
+	if rewriteModeenv {
+		if err := m.Write(); err != nil {
+			return err
+		}
+	}
+
+	defer func() {
```
Comment (mvo5) [inline boot/systems.go:382]: (nitpick) Given that the defer is added here and the next thing is the `return resealKeyToModeenv(dirs.GlobalRootDir, dev.Model(), m, expectReseal)` - would it make sense to simply write this as:
```
	const expectReseal = true
        if err := resealKeyToModeenv(dirs.GlobalRootDir, dev.Model(), m, expectReseal); err != nil {
            if cleanupErr := DropRecoverySystem(dev, systemLabel); cleanupErr != nil {
    			err = fmt.Errorf("%v (cleanup failed: %v)", err, cleanupErr)
            }
            return err
       }
       return nil
}
```
i.e. just avoid the slightly harder to read defer?

Or will this code change in the future and then the defer will be used for more than this one  failure case of "resealKeyToModeenv" ?

**Review by mvo5 — APPROVED** [review]
Thanks, this looks good, just some nitpick/suggestions inline

File: boot/systems.go (hunk 1 above)
Comment (pedronis) [reply boot/systems.go:368]: it was addressed, the original code in conditional stanza when I made the comment was returning an error in this case

File: boot/systems.go (hunk 3 above)
Comment (bboozzoo) [reply boot/systems.go:382]: Thanks for the suggestion, it looks nicer indeed.

**Pull request merged**